### ISSUE TITLE
question: is it possible to implement a 'suspend' feature?

### ISSUE THREAD

**Issue by jjess** [churchers/vm-bhyve#453]
I have read about how to suspend a vm with bhyve, for example here:

[https://github-wiki-see.page/m/FreeBSD-UPB/freebsd/wiki/Save-and-Restore-a-virtual-machine-using-bhyve](https://github-wiki-see.page/m/FreeBSD-UPB/freebsd/wiki/Save-and-Restore-a-virtual-machine-using-bhyve)

Is this kind of operation possible with vm-bhyve? maybe in the future?

Thanks in advance,

Jes

**Comment by churchers** [comment]
Yes it shouldn't take much to support the checkpoint/suspend features.

**Comment by syleishere** [comment]
Just tried it, but vm-bhyve does something funky with it taking it out of suspended state. Tried with 13.1 release branch as follows:
git clone --branch releng/13.1 https://git.FreeBSD.org/src.git /usr/src
cd /usr/src/sys/amd64/conf; cp GENERIC MYKERNEL; (add: options         BHYVE_SNAPSHOT)
cd /usr/src
make -j10 buildworld -DWITH_BHYVE_SNAPSHOT
make -j10 buildkernel KERNCONF=MYKERNEL
make installkernel KERNCONF=MYKERNEL
shutdown -r now
cd /usr/src; make installworld
shutdown -r now
etcupdate -B
-------------------------------

bhyvectl --suspend asterisk.ckp --vm=asterisk
(seems to store the files all in / directory)

now the funky part:
vm-bhyve wakes up the guest somehow. I do see originally the guest is suspended, but then vm-bhyve seems to just restart it as I try to connect to it with securecrt over ssh.

Need someone to look at this for sure.

**Comment by syleishere** [comment]
I'll post my asterisk.conf as well as I don't see there being an issue with it:

```
loader="uefi"
#graphics="yes"
xhci_mouse="yes"
cpu=8
cpu_sockets=2
cpu_cores=4
memory=8G
network0_type="virtio-net"
network0_switch="public"
#change to "nvme" if we get better server
disk0_type="virtio-blk"
disk0_name="disk0.img"
disk1_type="virtio-9p"
disk1_name="sharename=/mnt" 
disk1_dev="custom"
uuid="a5811c38-567e-11ed-bdeb-e4435bf65f00"
network0_mac="58:9c:fc:0c:4a:a6"
#please note: virtio-9p clients seem broken under FreeBSD 13.1, only FreeBSD 12 clients are reported working
```

**Comment by syleishere** [comment]
After playing with it more, I have suspend and restore working fine but not with 13.1 stable, only in 14 current. I didn't try it with vm-bhyve this time round but it does work, if interested you'll need to do following to upgrade to current:

```
git clone https://git.FreeBSD.org/src.git /usr/src
cd /usr/src/sys/amd64/conf
cp GENERIC-NODEBUG MYKERNEL (add: options         BHYVE_SNAPSHOT)
cd /usr/src
make -j12 buildworld -DWITH_BHYVE_SNAPSHOT -DWITH_MALLOC_PRODUCTION
make -j12 buildkernel KERNCONF=MYKERNEL
```

**Comment by syleishere** [comment]
FreeBSD 14:
`bhyvectl --suspend /vm/asterisk/default.ckp --vm=asterisk`
(suspends really quick like 10 seconds)
#now let's restore it:
```
bhyve -c 2 -m 4G -w -H -A \
        -s 0:0,hostbridge \
        -s 4:0,virtio-blk,/vm/asterisk/default.img  \
        -s 5:0,virtio-net,tap0 \
        -s 30,xhci,tablet \
        -s 31,lpc -l com1,stdio \
        -l bootrom,/usr/local/share/uefi-firmware/BHYVE_UEFI.fd \
        -r /vm/asterisk/default.ckp \
        asterisk
```

Instant restore. Quite happy with it, I'll stay with FreeBSD current with MALLOC_PRODUCTION flag, since netflix does that for its production servers anyways. There are some things that do not work properly like if you had an attached CD ISO, but for a very simple config like this of a simple FreeBSD guest using UEFI, works flawlessly.
